### PR TITLE
hw24-2: - done;

### PR DESCRIPTION
**Pull request by jah701**

### CODE REVIEW

File: src/main/java/internet/shop/dao/impl/ShoppingCartDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    public ShoppingCart create(ShoppingCart shoppingCart) {
+        Storage.addCart(shoppingCart);
+        return shoppingCart;
+    }
+
+    @Override
+    public Optional<ShoppingCart> get(Long id) {
+        return Storage.carts.stream()
+                .filter(x -> x.getId().equals(id)).findFirst();
```
Comment (matviichuk92) [inline src/main/java/internet/shop/dao/impl/ShoppingCartDaoImpl.java]: ментори завжди говорять, що "х" це не дуже інформативно

File: src/main/java/internet/shop/dao/impl/ShoppingCartDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            if (shoppingCart.getId().equals(item.getId())) {
+                return item.getProducts();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.carts.remove(get(id).orElseThrow());
```
Comment (matviichuk92) [inline src/main/java/internet/shop/dao/impl/ShoppingCartDaoImpl.java]: removeIf дуже крута фіча, можливо тобі сподобається більше

**Review by matviichuk92 — COMMENTED** [review]
хороша робота!

File: src/main/java/internet/shop/dao/ShoppingCartDao.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+public interface ShoppingCartDao {
+    ShoppingCart create(ShoppingCart shoppingCart);
+
+    Optional<ShoppingCart> get(Long id);
+
+    List<ShoppingCart> getAll();
+
+    boolean update(ShoppingCart shoppingCart);
+
+    List<Product> getProducts(ShoppingCart shoppingCart);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/ShoppingCartDao.java]: do we need this method on dao layer, can we just call shoppingCart.getProducts() ?

File: src/main/java/internet/shop/service/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public void clear(ShoppingCart shoppingCart) {
+        shoppingCart.clearProducts();
+    }
+
+    @Override
+    public ShoppingCart getByUserId(Long userId) {
+        return shoppingCartDao.get(userId).get();
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/service/impl/ShoppingCartServiceImpl.java]: you just pass userId into method, that excepts (Long cartId)

File: src/main/java/internet/shop/dao/impl/OrderDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    public Order create(Order order) {
+        Storage.addOrder(order);
+        return order;
+    }
+
+    @Override
+    public Optional<Order> get(Long id) {
+        return Storage.orders.stream()
+                .filter(x -> x.getId().equals(id)).findFirst();
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/OrderDaoImpl.java]: usually first letter of the object is used in naming inside of the stream operation (same goes to other methods)
```suggestion
                .filter(o -> o.getId().equals(id)).findFirst();
```

File: src/main/java/internet/shop/dao/impl/OrderDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            if (Storage.orders.get(i).getId().equals(order.getId())) {
+                Storage.orders.set(i, order);
+            }
+        }
+        return order;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.orders.remove(get(id - 1).orElseThrow());
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/OrderDaoImpl.java]: let's user removeIf here

File: src/main/java/internet/shop/dao/ShoppingCartDao.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.Optional;
+
+public interface ShoppingCartDao {
+    ShoppingCart create(ShoppingCart shoppingCart);
+
+    Optional<ShoppingCart> get(Long id);
+
+    List<ShoppingCart> getAll();
+
+    boolean update(ShoppingCart shoppingCart);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/ShoppingCartDao.java]:
```suggestion
    ShoppingCart update(ShoppingCart shoppingCart);
```

File: src/main/java/internet/shop/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+@Service
+public class OrderServiceImpl implements OrderService {
+    @Inject
+    private OrderDao orderDao;
+
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        Order order = new Order(shoppingCart.getId(),
+                shoppingCart.getProducts(), shoppingCart.getUserId());
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/service/impl/OrderServiceImpl.java]: don't ignore points from checklist 
https://mate-academy.github.io/jv-program-fulltime/02_jdbc_and_servlets/checklist/hw_14_2.html

File: src/main/java/internet/shop/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public ShoppingCart create(ShoppingCart shoppingCart) {
+        return shoppingCartDao.create(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCart addProduct(ShoppingCart shoppingCart, Product product) {
+        shoppingCart.getProducts().add(product);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/service/impl/ShoppingCartServiceImpl.java]: we need to call update method after, for shoppingCart in storage to receive changes, made here 
same for clear method

File: src/main/java/internet/shop/dao/impl/ProductDaoImpl.java [hunk 10]
```diff
@@ -10,34 +10,30 @@
...
-                .filter(x -> x.getId().equals(id))
+                .filter(o -> o.getId().equals(id))
                 .findFirst();
     }
 
     @Override
-    public boolean removeProduct(Long id) {
-        return Storage.products.remove(getProduct(id).orElseThrow());
+    public boolean remove(Long id) {
+        return Storage.products.remove(get(id).orElseThrow());
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ProductDaoImpl.java]: let's user here removeIf as well

File: src/main/java/internet/shop/dao/impl/ShoppingCartDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            if (Storage.carts.get(i).getId().equals(shoppingCart.getId())) {
+                Storage.carts.set(i, shoppingCart);
+            }
+        }
+        return shoppingCart;
+    }
+
+    @Override
+    public Optional<ShoppingCart> getByUserId(Long id) {
+        return Storage.carts.stream().filter(o -> o.getUserId().equals(id)).findFirst();
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ShoppingCartDaoImpl.java]:
```suggestion
        return Storage.carts.stream()
        .filter(o -> o.getUserId().equals(id))
        .findFirst();
```
don't forget about streams codestyle, same with others streams

File: src/main/java/internet/shop/dao/impl/UserDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+            if (Storage.users.get(i).getId().equals(user.getId())) {
+                Storage.users.set(i, user);
+            }
+        }
+        return user;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.users.remove(get(id).orElseThrow());
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/UserDaoImpl.java]: same here

File: src/main/java/internet/shop/service/ShoppingCartService.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+import internet.shop.model.Product;
+import internet.shop.model.ShoppingCart;
+
+public interface ShoppingCartService {
+    ShoppingCart create(ShoppingCart shoppingCart);
+
+    ShoppingCart addProduct(ShoppingCart shoppingCart, Product product);
+
+    ShoppingCart deleteProduct(ShoppingCart shoppingCart, Product product);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/service/ShoppingCartService.java]:
```suggestion
    boolean deleteProduct(ShoppingCart shoppingCart, Product product);
```
be more attentive with task description

**Comment by RomanDubovskyi** [comment]
fix this item from checklist, currently you don't clean your shoppingCart after purchase
-  In completeOrder() use List.copyOf(shoppingCart.getProducts()) to copy products from shoppingCart to order and don’t forget to clear ShoppingCart after you complete user’s order.

File: src/main/java/internet/shop/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -17,6 +17,7 @@
     public Order completeOrder(ShoppingCart shoppingCart) {
         Order order = new Order(shoppingCart.getId(),
                 List.copyOf(shoppingCart.getProducts()), shoppingCart.getUserId());
+        shoppingCart.clearProducts();
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/service/impl/OrderServiceImpl.java:20]:
```suggestion
        shoppingCartService.clear(shoppingCart);
```

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**